add ceph bucket notification adapter to knative

yuvalif opened the pull request:
Signed-off-by: Yuval Lifshitz <ylifshit@redhat.com>

This PR introduces a receive adapter between ceph bucket notifications and knative.
For more information on ceph bucket notifications see here: https://docs.ceph.com/docs/master/radosgw/notifications/

This is still WIP, should perform:
- [x] end to end testing on a cluster
- [x] add adapter unit tests

```release-note
NONE
```

matzew commented:
/ok-to-test

matzew commented on ceph/hack/test-pod.yaml:
```diff
@@ -0,0 +1,9 @@
+spec:
+  containers:
+    - name: curl-runner
+      image: docker.io/yuvalif/test-ceph-notifications
```
I'd recommend not doing this - let's use `ko` specific path, that other in kn land do too

matzew commented on ceph/Makefile:
```diff
@@ -0,0 +1,21 @@
+all: format vet adapter
```
we should not need a build file, let's use KO here

yuvalif replied:
removed the makefile

matzew commented:
let's change the `hack` folder to be named `sample`

matzew commented on ceph/samples/ceph-container-source.yaml:
```diff
@@ -0,0 +1,14 @@
+  labels:
+      app: ceph-source 
+spec:
+  image: docker.io/yuvalif/ceph-bucket-notification-source
```
replace the `docker.io` with `knative.dev/eventing-contrib/ceph/cmd/receive_adapter/`

matzew commented:
You may also add your code to here:

https://github.com/knative/eventing-contrib/blob/master/hack/update-deps.sh#L33

yuvalif commented:
> let's change the `hack` folder to be named `sample`

already have `sample` dir where I put sample deployment yaml. the `hack` dir is currently only for testing. maybe rename it to `test`?

matzew commented:
ok, let;s perhaps than keep `hack`, but let's remove the make file etc.

also not sure the `json` file is really needed ?  Or just for "tests" (aka sample?)

mattmoor commented on ceph/cmd/receive_adapter/main.go:
```diff
@@ -0,0 +1,24 @@
+package main
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2019 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2019 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2019 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2019 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main
```

mattmoor commented on ceph/pkg/adapter/adapter.go:
```diff
@@ -0,0 +1,87 @@
+package adapter
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2019 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package adapter
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2019 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package adapter
```

mattmoor commented on ceph/pkg/apis/v1alpha1/bucket_notification_types.go:
```diff
@@ -0,0 +1,63 @@
+package v1alpha1
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2019 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package v1alpha1
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2019 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package v1alpha1
```

yuvalif commented:
> ok, let;s perhaps than keep `hack`, but let's remove the make file etc.
> 
> also not sure the `json` file is really needed ? Or just for "tests" (aka sample?)

removed `hack` added the test files to the `sample` directory.
added more info on that into the readme

mattmoor commented on ceph/cmd/receive_adapter/main.go:
```diff
@@ -0,0 +1,24 @@
+
+	log.Printf("Sink is: %q. Listening on port: %q", *sink, *listenPort)
+
+	// if successfull, this should be a blocking call
```
<!--typo-bot-->
Found potential misspelling(s):
```suggestion
	// if successful, this should be a blocking call
```

mattmoor replied:
<!--typo-bot-->
Found potential misspelling(s):
```suggestion
	// if successful, this should be a blocking call
```

mattmoor commented on ceph/samples/ceph-container-source.yaml:
```diff
@@ -0,0 +1,14 @@
+metadata:
+  name: ceph-bucket-notification-source
+  labels:
+      app: ceph-source 
```
<!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
      app: ceph-source
```

mattmoor replied:
<!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
      app: ceph-source
```

yuvalif commented:
@googlebot I signed it!

yuvalif commented:
> @googlebot I signed it!

@googlebot I signed it!

matzew commented:
@googlebot I signed it!

/cc @yuvalif

matzew commented:
I mean: @yuvalif can you - again - do the @googlebot thing ? 

Sorry...

yuvalif commented:
@googlebot I signed it!

matzew commented:
/test pull-knative-eventing-contrib-integration-tests

mattmoor commented on ceph/pkg/adapter/adapter_test.go:
```diff
@@ -0,0 +1,133 @@
+		h.t.Error("Notification Body Mismatch")
+	}
+	w.WriteHeader(http.StatusOK)
+}
```
<!--whitespace-bot-->
Add trailing newline:
```suggestion
}

```

yuvalif replied:
`go fmt` does not like trailing newlines

aliok commented:
@matzew @yuvalif 
The CLA problem like that happened to me once as well. @matmoor's automated tasks either miss some info or that account didn't sign the CLA.
What I did was soft reset the branch and force push things so that PR don't end up having commits that are like "yuvalif and matmoor committed N days ago"

matzew commented:
Ah,  +1 on rebase and squash - good call, @aliok

yuvalif commented:
> @matzew @yuvalif
> The CLA problem like that happened to me once as well. @mattmoor's automated tasks either miss some info or that account didn't sign the CLA.
> What I did was soft reset the branch and force push things so that PR don't end up having commits that are like "yuvalif and mattmoor committed N days ago"

@aliok worked, thanks!

matzew commented:
/lgtm
